dali/benchmark/operator_bench.h

```diff
@@ -119,12 +119,15 @@ class OperatorBench : public DALIBenchmark {
     CUDA_CALL(cudaStreamSynchronize(0));
     for (auto _ : st) {
       op_ptr->Run(ws);
-      CUDA_CALL(cudaStreamSynchronize(0));
 
       int num_batches = st.iterations() + 1;
       st.counters["FPS"] = benchmark::Counter(batch_size * num_batches,
         benchmark::Counter::kIsRate);
     }
+
+    st.ResumeTiming();
```

Comment:
I have mixed feelings. On one hand, it tests only the kernel throughput without the synchronization overhead, on the other hand in real use cases usually a small number of kernels run one after another so there is always an overhead.
Maybe it should be parametrized?
Looking for other opinions.

Reply:
I see your point and I'm open to parametrizing it in some way. But just to clarify: the benchmark overhead (at least in case of Copy) added by the cudaStreamSynchronize API call itself was negligible, most of the overhead was caused by indirectly, by making Copy unable to schedule next memcpy while the previous one was still in progress. While I fully agree that in a real pipelines there are always some overheads, I doubt anything as bad as fully serializing the operators could happen in a pipeline.

Reply:
I might not understand the intentions completely, but shouldn't these two calls be swapped?
```c++
st.PauseTiming();
CUDA_CALL(cudaStreamSynchronize(0));
st.ResumeTiming();
```

Reply:
Wow, they *are* swapped on my local machine, no idea how this happened! I'm really sorry

Reply:
As you are not sure if this is the change you want, I think this PR needs to be redesigned to solve the problem while still keeping backward compatibility and flexibility. I would propose the following:

Adding extra parameter `unsigned pipelineLength` to `RunGPU` function in `operator_bench.h`, with a default value of `1` and with the following meaning: the benchmark would do its best to simulate a pipeline consisting of `pipelineLength` operators chained. In practice this would mean that in each benchmark iteration the operator would be run `pipelineLength` times, and then the CUDA stream will be synchronized.

This would keep things perfectly backward-compatible – for example running a benchmark with `Iterations(1000)` and `pipelineLength` defaulted to `1`, would run the operator 1000 times synchronizing the stream after each run. 

In my use case, when I would like to get rid of the synchronization overhead, I could run the benchmark with for example `Iterations(10)` and `pipelineLength` set to `100`, which would cause synchronizing the stream after each 100 benchmark runs.

What do you think about it? If that is OK, should we continue here or should I open a new PR?

Reply:
@hugo213 I agree with your change in general.

There's only one issue, the FPS counter is incorrect now, as it gets updated every iteration before the work was done, effectively timing only the scheduling of the work and not the work itself.

I suggest something like this:
```
    int64_t k = 0;
    for (auto _ : st) {
      op_ptr->Run(ws);
      if (k % update_fps_each_n == 0) {
        CUDA_CALL(cudaStreamSynchronize(0));
        int num_batches = st.iterations() + 1;
        st.counters["FPS"] = benchmark::Counter(batch_size * num_batches, benchmark::Counter::kIsRate); 
      }
      k++;
    }
    st.ResumeTiming();
    CUDA_CALL(cudaStreamSynchronize(0));
    st.PauseTiming();
    st.counters["FPS"] = benchmark::Counter(batch_size * st.iterations(), benchmark::Counter::kIsRate); 
    ```
    with `update_fps_after_n` being configurable.

Reply:
@jantonguirao you are right, this breaks the FPS counter, sorry for that.

I will wait for other reviewers opinion on whether they would prefer the original version, i.e. moving the synchronization out of the loop, or the new proposal with parametrizing it with `pipelineLength`. In the first case I will submit a commit with FPS fix as you suggested, and in the latter FPS counting won't be a problem.

Reply:
With the fixes, I think it is ok

Reply:
👍🏻 to what Albert wrote

Reply:
@jantonguirao Sorry for taking your time, but as I tried to implement your fix I realized that I have more questions about usage of Google Benchmark here and the idea of FPS counting in general:

1. *What exactly is FPS intended to be?* From what you wrote (`batch_size * st.iterations()`) it seems that you want it to be inverse of the time it takes to run **one iteration** of operator on one image. However, the FPS reported by the original `operator_bench.h` seem to be inverse of the time it takes to run **all iterations** of operator on one image.
2. *Is use of `.iterations()` in the original `operator_bench.h` valid?* The number of batches processed  is calculated in https://github.com/NVIDIA/DALI/blob/8af5e2b373be50a7d9f9bbdb54ae1d354f9595ed/dali/benchmark/operator_bench.h#L124 as `st.iterations() + 1`, however  `st.iterations()` seems to be `0` for the whole duration of the loop and be set to the number of iterations completed only after the loop. I tested this with a simple `cout`, but I can't find an exact place in Google Benchmark code where this happens. If it is true and `st.iterations()` is zero almost all of the time, then `num_batches` is always 1. This could mean that the author intended to count FPS as you wrote, but due to misunderstanding `st.iterations()` ended up counting all iterations as one.
3. *Is updating the FPS counter every iteration necessary?* Google Benchmark User Guide says (here: https://github.com/google/benchmark/blob/main/docs/user_guide.md#custom-counters) that "state.counters object is a std::map" which means that older values are simply overwritten. Wouldn't be setting the counter only once, in the end, to `batch_sizes * st.iterations()` enough?

Reply:
@hugo213 FPS here stands for "frames per second" (I think), and its value will be the total number of samples processed divided by the total time of the benchmark. `kIsRate` takes care of dividing the value by the time so we don't have to do it explicitly. I checked and it enough to report the custom counter once at the end of the test. So it'd be enough to do:

```
    int64_t sync_each_n = -1;  // make it configurable (?)
    int64_t batches = 0;
    for (auto _ : st) {
      op_ptr->Run(ws);
      if (sync_each_n > 0 && batches++ % sync_each_n == 0) {
        CUDA_CALL(cudaStreamSynchronize(0));
      }
    }
    st.ResumeTiming();
    CUDA_CALL(cudaStreamSynchronize(0));
    st.PauseTiming();
    st.counters["FPS"] = benchmark::Counter(batch_size * st.iterations(), benchmark::Counter::kIsRate); 
```

Note that I kept `sync_each_n` behavior in the loop. I am not sure if it'd be useful or not, but we can leave it off by default (`-1`) and allow the caller to override that.

Reply:
Thank you, I have pushed a commit adding the parameter.

PS. What I meant with the FPS question was what should for example happen if benchmark runs 1000 iterations of operator on 1 image in 1 second. The original (i.e. Nvidia's current upstream) benchmarks would then report 1FPS, and with the fix you suggested would report 1000FPS. I agree with your interpretation of what FPS should be, because as I wrote in question 2 current behaviour seems to be a result of a bug. The commit follows your interpretation and would report 1000FPS in the example above.

Reply:
The current version looks good. Thanks!